### PR TITLE
Hibernate 5 Multitenancy tutorial

### PR DESCRIPTION
**Pull request by joshiparthin**

### CODE REVIEW

File: hibernate5/pom.xml [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+			<artifactId>jboss-logging</artifactId>
+			<version>3.3.0.Final</version>
+		</dependency>
+		<dependency>
+			<groupId>mysql</groupId>
+			<artifactId>mysql-connector-java</artifactId>
+			<version>6.0.5</version>
+		</dependency>
+		<dependency>
+            <groupId>com.h2database</groupId>
```
Comment (KevinGilmore) [inline hibernate5/pom.xml]: Adjust the indentation here -- looks like this one uses spaces and the others use tabs. Spaces are preferred, but the main point is to be consistent.

File: hibernate5/pom.xml [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	<properties>
+		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+		<!-- Maven plugins -->
+        <maven-compiler-plugin.version>3.6.0</maven-compiler-plugin.version>
+	</properties>
+	<dependencies>
+		<dependency>
+			<groupId>org.hibernate</groupId>
+			<artifactId>hibernate-core</artifactId>
+			<version>5.2.6.Final</version>
```
Comment (KevinGilmore) [inline hibernate5/pom.xml]: Latest is 5.2.9.Final

File: hibernate5/pom.xml [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+		</dependency>
+		<dependency>
+			<groupId>mysql</groupId>
+			<artifactId>mysql-connector-java</artifactId>
+			<version>6.0.5</version>
+		</dependency>
+		<dependency>
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+            <version>1.4.193</version>
```
Comment (KevinGilmore) [inline hibernate5/pom.xml]: Latest is now 1.4.194

File: hibernate5/pom.xml [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+			<version>5.2.6.Final</version>
+		</dependency>
+		
+		<dependency>
+			<groupId>junit</groupId>
+			<artifactId>junit</artifactId>
+			<version>4.12</version>
+		</dependency>
+		<!-- https://mvnrepository.com/artifact/com.fasterxml/classmate -->
+		<dependency>
```
Comment (KevinGilmore) [inline hibernate5/pom.xml]: Are you certain that all of these other dependencies are needed in order to use hibernate? Are they not brought in as transitive properties when you include hibernate-core in your POM? Let me know what you find.

File: hibernate5/pom.xml (hunk 2 above)
Comment (joshiparthin) [inline hibernate5/pom.xml]: Done

File: hibernate5/pom.xml (hunk 4 above)
Comment (joshiparthin) [inline hibernate5/pom.xml]: Removed. Its working.

File: hibernate5/pom.xml (hunk 1 above)
Comment (joshiparthin) [inline hibernate5/pom.xml]: Done.

File: hibernate5/pom.xml (hunk 3 above)
Comment (joshiparthin) [inline hibernate5/pom.xml]: Done.

File: hibernate5/src/main/java/com/baeldung/hibernate/pojo/Supplier.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import javax.persistence.Table;
+
+import org.junit.runners.Suite.SuiteClasses;
+
+
+/**
+ * Suppliers generated by hbm2java
+ */
+@Entity(name = "Supplier")
+@Table(name ="Suppliers")
```
Comment (KevinGilmore) [inline hibernate5/src/main/java/com/baeldung/hibernate/pojo/Supplier.java]: Make table name "Supplier" also

File: pom.xml [hunk 6]
```diff
@@ -189,7 +189,7 @@
         <module>xstream</module>
 
         <module>struts2</module>
-
+        <module>hibernate5</module>
   </modules>
 
-</project>
```
Comment (KevinGilmore) [inline pom.xml]: There's a merge conflict with this file -- you'll need to update your branch from upstream, resolve the conflict(s),  and do another push. Not sure what the problem is exactly -- could be as simple as restoring the empty line at the end of the file.

File: hibernate5/pom.xml [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            <artifactId>hibernate-core</artifactId>
+            <version>5.2.9.Final</version>
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.12</version>
+        </dependency>
+        
+        <dependency>
```
Comment (KevinGilmore) [inline hibernate5/pom.xml]: Do we still need the mysql library?

File: hibernate5/src/main/java/com/baeldung/hibernate/App.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+ */
+public class App {
+    public static void main(String[] args) {
+        try {
+            System.out.println("Checking the system.");
+            SessionFactory sessionFactory = HibernateMultiTenantUtil.getSessionFactory();
+            Session currentSession = sessionFactory.withOptions().tenantIdentifier("h2db1").openSession();
+            Transaction transaction = currentSession.getTransaction();
+            transaction.begin();
+//            currentSession.createSQLQuery("DROP ALL OBJECTS;").executeUpdate();
```
Comment (KevinGilmore) [inline hibernate5/src/main/java/com/baeldung/hibernate/App.java]: if we don't need this line, let's remove it

File: hibernate5/src/main/java/com/baeldung/hibernate/config/Suppliers.hbm.xml [hunk 9]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+<!DOCTYPE hibernate-mapping PUBLIC "-//Hibernate/Hibernate Mapping DTD 3.0//EN"
+"http://www.hibernate.org/dtd/hibernate-mapping-3.0.dtd">
+<!-- Generated Feb 9, 2017 11:31:36 AM by Hibernate Tools 5.1.0.Final -->
+<hibernate-mapping>
+    <class name="com.baeldung.hibernate.pojo.Suppliers" table="suppliers" optimistic-lock="version">
```
Comment (KevinGilmore) [inline hibernate5/src/main/java/com/baeldung/hibernate/config/Suppliers.hbm.xml]: Shouldn't this use "Supplier" for the POJO and "supplier" for the table name?

File: hibernate5/src/main/java/hibernate.cfg.xml [hunk 10]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE hibernate-configuration PUBLIC
+		"-//Hibernate/Hibernate Configuration DTD 3.0//EN"
+		"http://www.hibernate.org/dtd/hibernate-configuration-3.0.dtd">
+<hibernate-configuration>
+    <session-factory>
+        <!--<property name="hibernate.connection.driver_class">com.mysql.jdbc.Driver</property> -->
```
Comment (KevinGilmore) [inline hibernate5/src/main/java/hibernate.cfg.xml]: We can remove the commented-out mysql lines.

File: hibernate5/src/main/java/hibernate.reveng.xml [hunk 11]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE hibernate-reverse-engineering PUBLIC "-//Hibernate/Hibernate Reverse Engineering DTD 3.0//EN" "http://hibernate.sourceforge.net/hibernate-reverse-engineering-3.0.dtd" >
+
+<hibernate-reverse-engineering>
+  <table-filter match-catalog="db1" match-name="suppliers"/>
```
Comment (KevinGilmore) [inline hibernate5/src/main/java/hibernate.reveng.xml]: supplier

File: pom.xml (hunk 6 above)
Comment (joshiparthin) [inline pom.xml]: @KevinGilmore  I think this is about parent pom.xml file. I've pulled, merged, resolved, commited and pushed. Let me know still if you are facing issues.

File: hibernate5/pom.xml (hunk 7 above)
Comment (joshiparthin) [inline hibernate5/pom.xml]: @KevinGilmore Done. Removed.

File: hibernate5/src/main/java/com/baeldung/hibernate/App.java (hunk 8 above)
Comment (joshiparthin) [inline hibernate5/src/main/java/com/baeldung/hibernate/App.java]: @KevinGilmore This class is not used in article. I have just taken here for testing the boostrapping of multitenancy. Let me know if you want me to delete it.

File: hibernate5/src/main/java/com/baeldung/hibernate/config/Suppliers.hbm.xml (hunk 9 above)
Comment (joshiparthin) [inline hibernate5/src/main/java/com/baeldung/hibernate/config/Suppliers.hbm.xml]: @KevinGilmore We have used entity classes mapping so this hbm is no more required. Earlier I tried it with hbm and thus this code was from that trial. We dont need this anymore and have removed in today's commits.

File: hibernate5/src/main/java/hibernate.cfg.xml (hunk 10 above)
Comment (joshiparthin) [inline hibernate5/src/main/java/hibernate.cfg.xml]: @KevinGilmore  Done.

File: hibernate5/src/main/java/hibernate.reveng.xml (hunk 11 above)
Comment (joshiparthin) [inline hibernate5/src/main/java/hibernate.reveng.xml]: @KevinGilmore as commented for hbm we dont need this file anymore and hence removed.

File: hibernate5/src/main/java/com/baeldung/hibernate/App.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+            currentSession.createCriteria(Suppliers.class).list().stream().forEach(System.out::println);
```
Comment (KevinGilmore) [inline hibernate5/src/main/java/com/baeldung/hibernate/App.java]: the createCriteria method is deprecated

File: hibernate5/src/main/java/com/baeldung/hibernate/MultiTenantHibernateTest.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+        }
+    }
+
+
+
+    private void initDb1(Session db1Session) {
+        System.out.println("Init DB1");
+        Transaction transaction = db1Session.getTransaction();
+        transaction.begin();
+        db1Session.createSQLQuery("DROP ALL OBJECTS").executeUpdate();
```
Comment (KevinGilmore) [inline hibernate5/src/main/java/com/baeldung/hibernate/MultiTenantHibernateTest.java]: the createSQLQuery method is deprecated

File: hibernate5/src/test/java/com/baeldung/hibernate/AppTest.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
+package com.baeldung.hibernate;
```
Comment (KevinGilmore) [inline hibernate5/src/test/java/com/baeldung/hibernate/AppTest.java]: Delete this class

File: hibernate5/src/main/java/com/baeldung/hibernate/MultiTenantHibernateTest.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
+package com.baeldung.hibernate;
```
Comment (KevinGilmore) [inline hibernate5/src/main/java/com/baeldung/hibernate/MultiTenantHibernateTest.java]: Let's move this class to src/test/java

File: hibernate5/src/main/java/com/baeldung/hibernate/HibernateMultiTenantUtil.java [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+//            sessionFactory = configuration.buildSessionFactory(serviceRegistry);
+            
+            
+        }
+        return sessionFactory;
+    }
+
+    private static SessionFactory makeSessionFactory(ServiceRegistry serviceRegistry) {
+        MetadataSources metadataSources = new MetadataSources( serviceRegistry );
+        for(Class annotatedClasses : getAnnotatedClasses()) {
```
Comment (KevinGilmore) [inline hibernate5/src/main/java/com/baeldung/hibernate/HibernateMultiTenantUtil.java]: I'm not fluent in Java 8 streaming API, but if there is way to leverage it here, let's do so.

Comment (joshiparthin) [inline hibernate5/src/main/java/com/baeldung/hibernate/HibernateMultiTenantUtil.java]: @KevinGilmore we can do that but will have to convert array results given by getAnnotedClasses() to list and then use stream. So I thought to keep it as it is. What do you feel ?

File: hibernate5/src/test/java/com/baeldung/hibernate/AppTest.java (hunk 14 above)
Comment (joshiparthin) [inline hibernate5/src/test/java/com/baeldung/hibernate/AppTest.java]: @KevinGilmore  done.

File: hibernate5/src/main/java/com/baeldung/hibernate/MultiTenantHibernateTest.java (hunk 13 above)
Comment (joshiparthin) [inline hibernate5/src/main/java/com/baeldung/hibernate/MultiTenantHibernateTest.java]: @KevinGilmore well since we are using legacy config method (SessionManager) for boostrapping against the EntityManager, we will have to use this functions though deprecated as of now for multitenancy example. As there is no documentation of how to do that with entitymanager bootstrapping method.

File: hibernate5/src/main/java/com/baeldung/hibernate/MultiTenantHibernateTest.java (hunk 15 above)
Comment (joshiparthin) [inline hibernate5/src/main/java/com/baeldung/hibernate/MultiTenantHibernateTest.java]: @KevinGilmore done.

File: hibernate5/src/main/java/com/baeldung/hibernate/MultiTenantHibernateTest.java (hunk 13 above)
Comment (KevinGilmore) [inline hibernate5/src/main/java/com/baeldung/hibernate/MultiTenantHibernateTest.java]: OK, thanks for the explanation.

File: hibernate5/src/main/java/com/baeldung/hibernate/HibernateMultiTenantUtil.java (hunk 16 above)
Comment (KevinGilmore) [inline hibernate5/src/main/java/com/baeldung/hibernate/HibernateMultiTenantUtil.java]: I don't feel strongly about it, but it may come up again in a higher-level review.

File: pom.xml [hunk 17]
```diff
@@ -225,7 +225,7 @@
                 <version>1.6.0</version>
                 <configuration>
                     <executable>maven</executable>
-
+        <module>hibernate5</module>
```
Comment (pivovarit) [inline pom.xml]: @KevinGilmore  @joshiparthin Why is this here?
